Lombiq.Hosting.Tenants.Maintenance/Readme.md

```diff
@@ -0,0 +1,49 @@
+# Lombiq Hosting - Tenant Maintenance for Orchard Core
+
+## About
+
+With the help of this module you can execute maintenance tasks on tenants.
+
+## Documentation
+
+Please see the below features for more information.
+
+### `Lombiq.Hosting.Tenants.Maintenance`
+
+This is the core functionality required to execute maintenance tasks on tenants. It is available on any tenant. To make your application execute maintenance tasks, you need to add the following to your `Startup.cs`:
+
+```csharp
+public void ConfigureServices(IServiceCollection services) =>
+    services.AddOrchardCms(
+        builder => builder.AddTenantFeatures(Lombiq.Hosting.Tenants.Maintenance.Constants.FeatureNames.Maintenance));
```

Comment:
Is this really needed? Enabled on all tenants? Same below.

Reply:
Well, you can disable maintenances with per-tenant configuration if the provider supports configuration. If not, then we can say that it's an important maintenance to be executed on all tenants. I'm curious what your thoughts are. In our cases the maintenances should be active on all tenants if applicable (e.g., a certain feature is active that needs to be fixed, or a URL replace). Do you think there's any scenario where it's not needed to be enabled on all tenants?

Reply:
We discussed that the URL replace should rather be something that only the Default tenant does. For something like user configuration, yes, but not necessarily all maintenance should be enabled on all tenants.

Reply:
Yes but that clearly must run only on the default shell which is a somewhat different scenario from having the option to run it on Tenant A, E, X, etc. as it always must be executed on the default shell; hence the .IsDefaultShell() condition in the ShouldExecuteAsync.

Reply:
Why just not enable that particular feature on all tenants? We don't want to run a NOOP on every tenant except the Default one.

Reply:
OK, I agree. However, now I have a related question.

Currently, I have two providers, one iterate through the shell settings and updates the tenant URL - it runs only on the default tenant. The other updates the site settings, but it is executed per tenant. Now if I want to do it on the default tenant, too, I'd need to iterate through each tenant, create a shell context, and execute a DB query with the site settings update which has multiple problems:

1. Performance issue.
2. We address tenants that might be totally inactive.
3. I met an annoying bug while using CreateScope for IShellHost during tenant activation, it causes an infinite loop during UI testing (it caused my literal headache in the past 1.5 weeks, and finally dropped using IShellHost entirely).
4. You don't even have a base URL set in the site settings for any of the representative tenants, and it's not even that important to set in my opinion.

So my suggestion is to either keep the site settings updater per tenant and run the tenant binding replacer on the default tenant; it means that those will be separate features. Or let's remove the site settings base URL updater because it's not that important.

What do you think?

Reply:
>So my suggestion is to either keep the site settings updater per tenant and run the tenant binding replacer on the default tenant; it means that those will be separate features.

Yep!